service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationControllerCitizen.kt

```diff
@@ -251,29 +262,36 @@ class ReservationControllerCitizen(
     }
 }
 
-data class PlacementDay(val shiftCare: Boolean, val contractDays: Boolean)
+data class PlacementDay(
+    val requiresReservation: Boolean,
+    val shiftCare: Boolean,
+    val contractDays: Boolean
+)
 
 private fun placementDay(
     date: LocalDate,
     isHoliday: Boolean,
-    placements: ReservationPlacements?
+    placements: List<ReservationPlacement>?,
+    backupPlacements: List<ReservationBackupPlacement>?,
+    contractDayRanges: Timeline?
 ): PlacementDay? {
-    if (placements == null) return null
+    val placement = placements?.find { it.range.includes(date) } ?: return null
 
     // Backup placements take precedence
-    val placement =
-        placements.backupPlacements.find { it.startDate <= date && date <= it.endDate }
-            ?: placements.regularPlacements.find { it.startDate <= date && date <= it.endDate }
+    val operationDays =
+        backupPlacements?.find { it.range.includes(date) }?.operationDays ?: placement.operationDays
 
-    return if (placement == null) {
-        null
-    } else if (placement.operationDays == setOf(1, 2, 3, 4, 5, 6, 7)) {
-        PlacementDay(shiftCare = true, contractDays = placement.hasContractDays)
-    } else if (placement.operationDays.contains(date.dayOfWeek.value) && !isHoliday) {
-        PlacementDay(shiftCare = false, contractDays = placement.hasContractDays)
-    } else {
-        null
-    }
+    val contractDays = contractDayRanges?.includes(date) ?: false
+
+    val shiftCare = operationDays == setOf(1, 2, 3, 4, 5, 6, 7)
+    val isOperationDay = operationDays.contains(date.dayOfWeek.value)
+
+    return PlacementDay(
+            requiresReservation = placement.type.requiresAttendanceReservations(),
+            shiftCare = shiftCare,
+            contractDays = contractDays
+        )
+        .takeIf { shiftCare || isOperationDay && !isHoliday }
```

Comment:
nitpick: I prefer adding explicit parenthesis as at least I never remember order of precedence with AND/OR